pennylane_lightning/src/util/Util.hpp

```diff
@@ -204,89 +207,465 @@ template <class T> inline auto dimSize(const std::vector<T> &data) -> size_t {
     return static_cast<size_t>(log2(s_sqrt));
 }
 
+/**
+ * @brief Calculates the inner-product using OpenMP.
+ *
+ * @tparam T Floating point precision type.
+ * @param v1 Complex data array 1.
+ * @param v2 Complex data array 2.
+ * @param result Calculated inner-product of v1 and v2.
+ * @param data_size Size of data arrays.
+ */
+template <class T>
+inline static void
+omp_innerProd(const std::complex<T> *v1, const std::complex<T> *v2,
+              std::complex<T> &result, const size_t data_size) {
+#if defined _OPENMP
+#pragma omp declare \
+            reduction (sm:std::complex<T>:omp_out=ConstSum(omp_out, omp_in)) \
+            initializer(omp_priv=std::complex<T> {0, 0})
+#endif
+
+#if defined _OPENMP
+#pragma omp parallel for default(none) shared(v1, v2, data_size)               \
+    reduction(sm                                                               \
+              : result)
+#endif
+    for (size_t i = 0; i < data_size; i++) {
+        result = ConstSum(result, ConstMult(*(v1 + i), *(v2 + i)));
+    }
+}
+
 /**
  * @brief Calculates the inner-product using the best available method.
  *
  * @tparam T Floating point precision type.
- * @param data_1 Complex data array 1.
- * @param data_2 Complex data array 2.
+ * @param v1 Complex data array 1.
+ * @param v2 Complex data array 2.
  * @param data_size Size of data arrays.
  * @return std::complex<T> Result of inner product operation.
  */
 template <class T>
-auto innerProd(const std::complex<T> *data_1, const std::complex<T> *data_2,
-               const size_t data_size) -> std::complex<T> {
+inline auto innerProd(const std::complex<T> *v1, const std::complex<T> *v2,
+                      const size_t data_size) -> std::complex<T> {
     std::complex<T> result(0, 0);
 
-    if constexpr (USE_CBLAS) {
-        if constexpr (std::is_same_v<T, float>) {
-            cblas_cdotu_sub(data_size, data_1, 1, data_2, 1, &result);
-        } else if constexpr (std::is_same_v<T, double>) {
-            cblas_zdotu_sub(data_size, data_1, 1, data_2, 1, &result);
-        }
-    } else {
+#if defined(USE_CBLAS) && USE_CBLAS
+    if constexpr (std::is_same_v<T, float>) {
+        cblas_cdotu_sub(data_size, v1, 1, v2, 1, &result);
+    } else if constexpr (std::is_same_v<T, double>) {
+        cblas_zdotu_sub(data_size, v1, 1, v2, 1, &result);
+    }
+#else
+    if (data_size > DOTU_STD_CROSSOVER) {
```

Comment:
Should this be 
```suggestion
    if (data_size < DOTU_STD_CROSSOVER) {
```
Or does OMP get enabled for small sizes only?